cmd/cerbosctl/hub/epdp/list_candidates_test.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2021-2024 Zenauth Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+//go:build tests
+
+package epdp
+
+import (
+	"context"
+	"fmt"
+	"io/fs"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/cerbos/cerbos/internal/policy"
+	"github.com/cerbos/cerbos/internal/test"
+)
+
+func TestListCandidates(t *testing.T) {
+	testCases := []struct {
+		testFile     string
+		files        []string
+		expectedList []string
+	}{
+		{
+			testFile: "all_have_metadata.txt",
+			files: []string{
+				"purchase_order.yaml",
+				"regional/purchase_order.yaml",
+				"regional/uk/purchase_order.yaml",
+			},
+			expectedList: []string{
+				"resource.purchase_order.vdefault",
+				"resource.purchase_order.vdefault/regional",
+				"resource.purchase_order.vdefault/regional.uk",
+			},
+		},
+		{
+			testFile: "ancestor_not_in_repo.txt",
+			files: []string{
+				"regional/purchase_order.yaml",
+				"regional/uk/purchase_order.yaml",
+			},
+			expectedList: []string{
+				"resource.purchase_order.vdefault/regional",
+			},
+		},
+		{
+			testFile: "ancestors.txt",
+			files: []string{
+				"purchase_order.yaml",
+				"regional/purchase_order.yaml",
+				"regional/uk/purchase_order.yaml",
+			},
+			expectedList: []string{
+				"resource.purchase_order.vdefault",
+				"resource.purchase_order.vdefault/regional",
+			},
+		},
+		{
+			testFile: "ancestors_not_in_repo.txt",
+			files: []string{
+				"regional/uk/purchase_order.yaml",
+			},
+			expectedList: []string{
+				"resource.purchase_order.vdefault/regional.uk",
+			},
+		},
+		{
+			testFile: "descendant_observed_first.txt",
+			files: []string{
+				"regional/purchase_order.yaml",
+				"purchase_order.yaml",
+			},
+			expectedList: []string{
+				"resource.purchase_order.vdefault",
+				"resource.purchase_order.vdefault/regional",
+			},
+		},
+		{
+			testFile: "irrelevant_metadata.txt",
+			files: []string{
+				"purchase_order.yaml",
+				"regional/purchase_order.yaml",
+				"regional/uk/purchase_order.yaml",
+			},
+			expectedList: []string{
+				"resource.purchase_order.vdefault",
+				"resource.purchase_order.vdefault/regional",
+				"resource.purchase_order.vdefault/regional.uk",
+			},
+		},
+		{
+			testFile: "no_metadata.txt",
+			files: []string{
+				"purchase_order.yaml",
+				"regional/purchase_order.yaml",
+				"regional/uk/purchase_order.yaml",
+			},
+			expectedList: []string{
+				"resource.purchase_order.vdefault",
+				"resource.purchase_order.vdefault/regional",
+				"resource.purchase_order.vdefault/regional.uk",
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.testFile, func(t *testing.T) {
+			fsys := test.ExtractTxtArchiveToFS(t, filepath.Join("testdata", testCase.testFile))
+			pl := policyLoader{fsys: fsys}
+
+			candidates, err := listCandidates(context.Background(), pl.loadPolicy, testCase.files...)
```

Comment:
This test is not good because `listCandidates` is missing the crucial `index.Build()` part that would make most of your test cases fail because the repos are invalid. So, it's not behaving as it would in a real-world scenario.